File: problem_049.c
// -*- compile-command: "gcc -o problem_049 problem_049.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "ceuler.h"

int main()
{
    printf("Project Euler - Problem 49:\n"
           "What 12 digit sequence do you form by concatenating three four-digit primes that are permutations of each other and increase by the same amount?\n\n");

    int basePrime = 1000, prime2, prime3;
    int found = 0;

    while (basePrime++ < 10000)
    {
        if (isPrime(basePrime) == false) continue;

        //printf("%d, %ud\n", basePrime, numSignature(basePrime)); // DEBUG
        int max = 10000 - basePrime / 2;
        int diff = 0;
        for (diff=1; diff<max; diff++)
        {
            prime2 = basePrime + diff;
            prime3 = basePrime + (2 * diff);

            if (isPrime(prime2) == false || isPrime(prime3) == false) continue;
            if (prime2 > 9999 || prime3 > 9999) continue;
            if (numSignature(basePrime) != numSignature(prime2)) continue;
            if (numSignature(basePrime) != numSignature(prime3)) continue;

            //printf("%d, %d, %d\n", basePrime, prime2, prime3);
            found++;
            break;
        }
        if (found == 2) break;
    }

    printf("12 digit number: %d%d%d\n", basePrime, prime2, prime3);
    return 0;
}

File: problem_049.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "swiftc problem_049.swift -o problem_049" -*-
// Copyright (c) 2016 Michael Caldwell
import Cocoa

print("Project Euler - Problem 49:")
print("What 12 digit sequence do you form by concatenating three four-digit primes that are permutations of each other and increase by the same amount?\n")

func isPrime(n:Int) -> Bool {
    if n < 2 { return false }
    if n == 2 { return true }
    if n % 2 == 0 { return false }

    for var i = 3; i * i <= n ; i+=2 {
        if n % i == 0 {
            return false
        }
    }

    return true;
}

func getSig(var a:Int) -> [Int] {
    var aa = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

    while a > 0 {
        aa[a%10] += 1
        a /= 10
    }

    return aa
}

var primes:[Int] = []
var signature:[[Int]] = []

for n in 1488...9999 {
    if isPrime(n) == true {
        primes.append(n)
        signature.append(getSig(n))
    }
}

var answer = "148748178147"

for a in 0...primes.count-1 {
    for b in a+1...primes.count-1 {
        if primes[a] + primes[b] > 9999 {
            break
        }

        if signature[a] == signature[b] {
            var c = primes[b] * 2 - primes[a]
            if primes.contains(c) {
                c = primes.indexOf(c)!
                if signature[a] == signature[c] {
                    answer = String(primes[a])+String(primes[b])+String(primes[c])
                    break
                }
            }
        }
    }
    if answer != "148748178147" {
        break
    }
}

print(answer)
